crates/bevy_transform/src/systems.rs

```diff
@@ -1,75 +1,84 @@
-use bevy_ecs::prelude::{Changed, Entity, Query, With, Without};
+use bevy_ecs::prelude::{Changed, Entity, Or, Query, With, Without};
 use bevy_hierarchy::{Children, Parent};
 
+/// Update [`GlobalTransform`] component of entities that aren't in the hierarchy
+pub fn transform_propagate_system_flat(
+    mut query: Query<
+        (&Transform, &mut GlobalTransform),
+        (Changed<Transform>, Without<Parent>, Without<Children>),
+    >,
+) {
+    for (transform, mut global_transform) in query.iter_mut() {
+        *global_transform = GlobalTransform::from(*transform);
+    }
+}
+
 /// Update [`GlobalTransform`] component of entities based on entity hierarchy and
 /// [`Transform`] component.
 pub fn transform_propagate_system(
     mut root_query: Query<
         (
-            Option<(&Children, Changed<Children>)>,
+            Entity,
+            &Children,
             &Transform,
-            Changed<Transform>,
+            Or<(Changed<Children>, Changed<Transform>)>,
             &mut GlobalTransform,
-            Entity,
         ),
         Without<Parent>,
     >,
-    mut transform_query: Query<(
-        &Transform,
-        Changed<Transform>,
-        &mut GlobalTransform,
-        &Parent,
-    )>,
+    transform_query: Query<(&Transform, Changed<Transform>, &mut GlobalTransform), With<Parent>>,
+    parent_query: Query<&Parent>,
     children_query: Query<(&Children, Changed<Children>), (With<Parent>, With<GlobalTransform>)>,
 ) {
-    for (children, transform, transform_changed, mut global_transform, entity) in
-        root_query.iter_mut()
-    {
-        let mut changed = transform_changed;
-        if transform_changed {
-            *global_transform = GlobalTransform::from(*transform);
-        }
+    root_query.par_for_each_mut(
+        1,
+        |(entity, children, transform, changed, mut global_transform)| {
+            if changed {
+                *global_transform = GlobalTransform::from(*transform);
+            }
 
-        if let Some((children, changed_children)) = children {
-            // If our `Children` has changed, we need to recalculate everything below us
-            changed |= changed_children;
             for child in children.iter() {
                 let _ = propagate_recursive(
                     &global_transform,
-                    &mut transform_query,
+                    &transform_query,
+                    &parent_query,
                     &children_query,
-                    *child,
                     entity,
+                    *child,
                     changed,
                 );
             }
-        }
-    }
+        },
+    );
 }
 
 fn propagate_recursive(
     parent: &GlobalTransform,
-    transform_query: &mut Query<(
-        &Transform,
-        Changed<Transform>,
-        &mut GlobalTransform,
-        &Parent,
-    )>,
+    transform_query: &Query<(&Transform, Changed<Transform>, &mut GlobalTransform), With<Parent>>,
```

Comment:
I wonder if we should rename this to something like `unsafe_transform_query` here (and above). The requirements for this to be safely accessed are quite subtle - even `Query::get` is probably UB. 
This seems like the kind of code which would be blindly copied for user's custom hierarchy, so we can at least nudge the gun away from the foot-direction with wording here.

We really need `UnsafeCell` equivalent queries - they're on the list.

Reply:
Could/should we make this system generic? - that would avoid the copy/paste problem

We couldn't allow arbitrary hierarchies, since the `Parent` equivalent component could have a malicious `Deref` impl to make the check always pass. However, I see no reason that this couldn't access arbitrary other components.

Reply:
This is done in #4216, which we can merge before or after this PR.